test/e2e/framework/service_util.go

```diff
@@ -1450,3 +1441,99 @@ func GetServiceLoadBalancerCreationTimeout(cs clientset.Interface) time.Duration
 	}
 	return LoadBalancerCreateTimeoutDefault
 }
+
+// TestAffinityTransition function tests whether the service affinity works a
+// expected. If affinity is expected, the test will return true once
+// AffinityConfirmCount number of same response observed in a row. If affinity
+// is not expected, the test will keep observe until different responses
+// observed. The function will return false only when no expected responses
+// observed before timeout.
+func TestAffinityTransition(jig *ServiceTestJig, execPod *v1.Pod, targetIp string, targetPort int, shouldHold bool) bool {
+	var hostname, response string
+	targetIpPort := net.JoinHostPort(targetIp, strconv.Itoa(targetPort))
+	cmd := fmt.Sprintf(`wget -qO- http://%s/ -T 2`, targetIpPort)
+	affinityCount := 0
+	timeout := ServiceTestTimeout
+	if execPod == nil {
+		timeout = LoadBalancerPollTimeout
+	}
+	if pollErr := wait.PollImmediate(Poll, timeout, func() (bool, error) {
+		if execPod != nil {
+			if stdout, err := RunHostCmd(execPod.Namespace, execPod.Name, cmd); err != nil {
+				Logf("Failed to get response from %s. Retry until timeout", targetIpPort)
+				hostname = ""
+				affinityCount = 0
+				return false, nil
+			} else {
+				response = stdout
+			}
+		} else {
+			rawResponse := jig.GetHTTPContent(targetIp, targetPort, timeout, "")
+			response = rawResponse.String()
+		}
+		Logf("Got pod response: %s", response)
+		if shouldHold {
+			if hostname == response {
+				affinityCount++
+			} else {
+				affinityCount = 0
+				hostname = response
+			}
+			return affinityCount >= AffinityConfirmCount, nil
+		} else {
+			if hostname == "" {
+				hostname = response
+				return false, nil
+			} else {
+				return hostname != response, nil
+			}
+		}
+	}); pollErr != nil {
+		// Affinity check timed out.
+		return false
+	}
+	return true
+}
+
+// TestAffinity tests if session affinity works as expecterd. The service
+// should not be in a transition state when calling this function.
+func TestAffinity(jig *ServiceTestJig, execPod *v1.Pod, targetIp string, targetPort int, shouldHold bool) bool {
+	var hostname, response string
+	targetIpPort := net.JoinHostPort(targetIp, strconv.Itoa(targetPort))
+	cmd := fmt.Sprintf(`wget -qO- http://%s/ -T 2`, targetIpPort)
+	isHold := true
+	affinityCount := 0
+	timeout := ServiceTestTimeout
+	if execPod == nil {
+		timeout = LoadBalancerPollTimeout
+	}
+	if pollErr := wait.PollImmediate(Poll, timeout, func() (bool, error) {
+		if execPod != nil {
+			if stdout, err := RunHostCmd(execPod.Namespace, execPod.Name, cmd); err != nil {
+				Logf("Failed to get response from %s. Retry until timeout", targetIpPort)
+				affinityCount = 0
+				return false, nil
+			} else {
+				response = stdout
+			}
+		} else {
+			rawResponse := jig.GetHTTPContent(targetIp, targetPort, timeout, "")
+			response = rawResponse.String()
+		}
+		Logf("Got pod response: %s", response)
```

Comment:
Can we do this to avoid the logic repeated:
```
type affinityTracker struct {
  hostTrace []string
}

func (at *affinityTracker) record(host string) {
  at.hostTrace = append(at.hostTrace, host)
}

func (at *affinityTracker) check(count int) {
  if len(at.hostTrace) < count {
    return false
  }
  last := at.hostTrace[len(at.hostTrace)-count:]
  host := at.hostTrace[len(at.hostTrace)-count]
  for _, h := range last {
    if h != host {
      return false
    }
  }
  return true
}
```

Reply:
You can then use it in both tests

Reply:
Done. Take a look.